### PR TITLE
Add \a escape sequence

### PR DESCRIPTION
**Pull request by r00ster91**
The escape sequence that beeps/bells.

Right now when you do
```Crystal
puts "\x07"
```
then you can hear a beep/bell and it already works because `07` is the HEX value in ASCII of `\a`.
But I still think that we should add the escape sequence `\a` directly because it's really some kind of special feature. Thats pretty much the only thing you can hear in a terminal.

This escape sequence can be useful for example when you are making an timer or something like that. So you can notify the user while he's doing something else. And it can be also helpful at debugging something without a display. Or generally debugging.

### CODE REVIEW

**Comment by oprypin** [comment]
I think when the original developer was deciding for the first time what escape sequences to include, they made an educated decision on which ones are important. I don't think it's a matter of forgetting to include it or anything like that. More is not always better. Ruby having it is not a good argument.

**Comment by oprypin** [comment]
Now that I look at it more carefully, this does seem like an unusual omission, because Crystal has all other escape sequences from these two

- https://en.wikipedia.org/wiki/Escape_sequences_in_C#Table_of_escape_sequences
- https://docs.python.org/3.6/reference/lexical_analysis.html#index-18

File: spec/compiler/lexer/lexer_spec.cr [hunk 1]
```diff
@@ -230,6 +230,7 @@ describe "Lexer" do
   it_lexes_number :i8, ["0i8", "0"]
 
   it_lexes_char "'a'", 'a'
+  it_lexes_char "'\\a'", '\a'
```
Comment (oprypin) [inline spec/compiler/lexer/lexer_spec.cr]: I think to keep release compatibility you would have to write `7.chr` not `'\a'`

File: src/compiler/crystal/syntax/lexer.cr [hunk 2]
```diff
@@ -493,6 +493,8 @@ module Crystal
             case char
             when '\\'
               case char = next_char
+              when 'a'
+                io << "\x07"
```
Comment (Sija) [inline src/compiler/crystal/syntax/lexer.cr]: I'd keep `"\u{7}"` format.

File: src/compiler/crystal/syntax/lexer.cr [hunk 3]
```diff
@@ -1789,6 +1793,8 @@ module Crystal
             end
           else
             case char = next_char
+            when 'a'
+              string_token_escape_value "\x07"
```
Comment (Sija) [inline src/compiler/crystal/syntax/lexer.cr]: ditto

File: src/compiler/crystal/syntax/lexer.cr (hunk 2 above)
Comment (oprypin) [reply src/compiler/crystal/syntax/lexer.cr]: Yeah basically you have to do whatever the `\b` addition does. New source code must still compile with the old compiler.

**Comment by r00ster91** [comment]
Ok now I did whatever the `\b` addition does. In every file. I searched with sublime through every single file for `\b`.
But CircleCI says:
```Crystal
Syntax error in src/string.cr:3942: invalid char escape sequence



      when '\a' then io << "\\a"

           ^

**Comment by oprypin** [comment]
Can't use `\a` anywhere, except when it's `\\a`

**Comment by oprypin** [comment]
https://github.com/crystal-lang/crystal/commit/4f462bae0264c126854e20103f4379f2182c5277#diff-627df40ff6c553c599dce26c4eeb3fb3R406

**Comment by r00ster91** [comment]
Hmm now its giving weird these weird errors
```Crystal
Error: formatting 'spec/std/json/lexer_spec.cr' produced changes

Error: formatting 'spec/std/json/serialization_spec.cr' produced changes

Error: formatting 'spec/std/char_spec.cr' produced changes

Error: formatting 'spec/std/string_spec.cr' produced changes

Error: formatting 'src/json/lexer.cr' produced changes

Error: formatting 'src/json/builder.cr' produced changes

Error: formatting 'src/string.cr' produced changes
```

**Comment by oprypin** [comment]
Run `crystal tool format`

**Comment by RX14** [comment]
I don't see the usecase. All unprintable characters shouldn't be escale sequences.

**Comment by r00ster91** [comment]
Okay then im closing this

**Comment by Sija** [comment]
@RX14 `\b` is unprintable as well, I see no reason to skip this one, while having all others included. This should be added only if for completeness sake.

**Comment by j8r** [comment]
Agree, there are also some languages that support it like Ruby and Python, and others don't like Rust and Go AFAIK.

**Comment by straight-shoota** [comment]
Go supports `\a`, Rust does not.

**Comment by RX14** [comment]
@Sija we should remove `\b`.

**Comment by RX14** [comment]
It seems rust supports only `\u` `\x`, and then `\n`, `\r`, `\t`, and `\0`. And obviously `\\`. We should duplicate that.

**Comment by Sija** [comment]
@RX14 or add `\a`...

**Comment by RX14** [comment]
`\a` is platform-specific and obsolete, so is `\b`.

**Comment by straight-shoota** [comment]
`\b\f\v\e` are actually recognized (see https://crystal-lang.org/docs/syntax_and_semantics/literals/string.html ). I think we all agree we should either remove all of them or add `\a`. I don't have a strong position on this. They are probably mostly useless - but on the other hand it doesn't really hurt to have them and in rare instances it might help interoprability.

@RX14 How is `\a` platform-specific? It's just an alternative name to express codepoint 7.

**Comment by RX14** [comment]
I absolutely agree we should add all of them or remove all of them except r, n, t. I just think we should remove complexity, and arcane, typically unused escape sequences.

**Comment by straight-shoota** [comment]
I'd say `\e` should also stay. It is even used  in Colorize and Benchmark::IPS

**Comment by Sija** [comment]
@RX14 What's the point of removing well-known functionalities (even if they indeed have _arcane_ heritage), just for removal sake? Wouldn't be better to stay compatible with other languages and stay _modern_ in ways that are meaningful?

**Comment by asterite** [comment]
I didn't include "\a" because I don't find it useful. I wouldn't mind having it in the language, though.

**Comment by RX14** [comment]
I don't care enough to bikeshed over it. If this many people feel it should be in the language, then just add it and I won't use it.

**Comment by asterite** [comment]
(if someone approves this and wants to merge this, please squash :-)

**Comment by Sija** [comment]
Also, json additions need to be removed, since `\a` is invalid character in JSON string.

**Comment by Sija** [comment]
@r00ster91 FYI, there's still one left in `src/json/builder.cr`.

File: spec/std/string_spec.cr [hunk 4]
```diff
@@ -1444,6 +1444,7 @@ describe "String" do
     "a".dump.should eq %("a")
     "\\".dump.should eq %("\\\\")
     "\"".dump.should eq %("\\\"")
+    # TODO: "\a".dump.should eq %("\\a")
```
Comment (straight-shoota) [inline spec/std/string_spec.cr]: Maybe add a note `uncomment after 0.24.2`?

File: src/compiler/crystal/syntax/lexer.cr [hunk 5]
```diff
@@ -493,6 +493,8 @@ module Crystal
             case char
             when '\\'
               case char = next_char
+              when 'a'
+                io << "\u{7}"
```
Comment (straight-shoota) [inline src/compiler/crystal/syntax/lexer.cr]: Is there a reason these were not touched by #5882 @sija?

File: spec/std/string_spec.cr (hunk 4 above)
Comment (r00ster91) [reply spec/std/string_spec.cr]: I think this is enough. Right after 0.24.2 gets released, I will finish this with another PR and there I will just look here to see what I need to do.

File: src/compiler/crystal/syntax/lexer.cr (hunk 5 above)
Comment (Sija) [reply src/compiler/crystal/syntax/lexer.cr]: Nope, just an omission.

Comment (r00ster91) [reply src/compiler/crystal/syntax/lexer.cr]: Should I make a PR to also make these strings chars?

Comment (Sija) [reply src/compiler/crystal/syntax/lexer.cr]: I'd suggest to merge this one first.

**Pull request merged**